jws/jws-mediator/src/main/java/org/wso2/financial/services/apim/mediation/policies/jws/JwsRequestHeaderValidationMediator.java

```diff
@@ -0,0 +1,596 @@
+                log.error(errorDescription);
+                setErrorResponseProperties(messageContext, "Bad Request", errorDescription, "400");
+                throw new SynapseException(errorDescription);
+            }
+
+            return true;
+        } catch (IOException | ParseException | JOSEException | SynapseException e) {
+            log.error(e);
+            setErrorResponseProperties(messageContext, "Bad Request", e.getMessage(), "400");
+            throw new SynapseException(e.getMessage());
+        }
+    }
+
+    /**
+     * Validates a detached JWS signature against the provided payload and JWKS URL.
+     *
+     * @param messageContext The Synapse MessageContext
+     * @param detachedJWS The detached JWS signature in the format <header>..<signature>
+     * @param payload The payload to verify against the detached JWS signature
+     * @param jwksUrl The URL of the JWKS endpoint to retrieve the public key for verification
+     * @return true if the detached JWS signature is valid, false otherwise
+     * @throws ParseException
+     * @throws JOSEException
+     * @throws IOException
+     */
+    protected boolean validateDetachedJWS(MessageContext messageContext, String detachedJWS, String payload,
+                                          String jwksUrl) throws ParseException, JOSEException, IOException {
+
+        String[] parts = detachedJWS.split("\\.");
+        if (parts.length != 3) {
+            throw new IllegalArgumentException("Invalid detached JWS format. Expected format: <header>..<signature>");
+        }
+
+        Base64URL encodedHeader = new Base64URL(parts[0]);
+        Base64URL encodedSignature = new Base64URL(parts[2]);
+
+        // Parse JWS header
+        JWSHeader jwsHeader = JWSHeader.parse(encodedHeader);
+
+        PublicKey publicKey = getPublicKeyFromJWKS(jwksUrl, jwsHeader.getKeyID());
+
+        // Create custom critical header policy
+        Set<String> deferredCritHeaders = JwsUtils.getDifferedCritHeaders();
+
+        boolean areCustomClaimsValid = validateCustomClaims(messageContext, jwsHeader, jwksUrl);
+
+        if (!areCustomClaimsValid) {
+            return false;
+        }
+
+        JWSVerifier verifier = JwsUtils.getJwsVerifier(publicKey, jwsHeader, deferredCritHeaders);
+
+        if (JwsUtils.isPayloadB64Encoded(jwsHeader)) {
+            // b64=true
+            String reconstructedJws = JwsUtils.reconstructJws(detachedJWS, payload);
+            JWSObject jwsObject = JWSObject.parse(reconstructedJws);
+            return jwsObject.verify(verifier);
+        } else {
+            // b64=false (raw payload used in signature input)
+            // Construct the signing input: base64url(header) + "." + payload
+            String signingInput = encodedHeader + "." + payload;
+            byte[] signingInputBytes = signingInput.getBytes(StandardCharsets.US_ASCII);
+            return verifier.verify(jwsHeader, signingInputBytes, encodedSignature);
+        }
```

Comment:
Shall we add few debug logs

Reply:
Fixed with 55d7e83f22cd8b61c36a2ec94553099c99069c10